### PR TITLE
Minor: more flexible pool size setting for datafusion-cli

### PR DESCRIPTION
**Pull request by yjshen**
## Which issue does this PR close?

re #7424.

## Rationale for this change

Find the memory pool size setting string in datafusion-cli during performance testing, and we could make the byte string more flexible.

## What changes are included in this PR?

Parse the pool byte string as the number of bytes.

## Are these changes tested?

Add unit test and tested locally.

## Are there any user-facing changes?

No.

### CODE REVIEW

File: datafusion-cli/src/main.rs [hunk 1]
```diff
@@ -252,26 +253,128 @@ fn is_valid_memory_pool_size(size: &str) -> Result<(), String> {
...
+        assert_conversion("4kbx", Err("Invalid memory pool size '4kbx'".to_string()));
+        assert_conversion(
+            "-20mb",
+            Err("Negative memory pool size value is not allowed '-20mb'".to_string()),
+        );
+        assert_conversion(
+            "-100",
+            Err("Negative memory pool size value is not allowed '-100'".to_string()),
+        );
+
```
Comment (Weijun-H) [inline datafusion-cli/src/main.rs:377]:
```suggestion
        assert_conversion(
            "12k12k",
            Err("Invalid memory pool size '12k12k'".to_string()),
        );
```

Comment (Weijun-H) [reply datafusion-cli/src/main.rs:377]: This case should have failed, but it passed.

File: datafusion-cli/src/main.rs [hunk 2]
```diff
@@ -252,26 +253,128 @@ fn is_valid_memory_pool_size(size: &str) -> Result<(), String> {
...
+    fn multiplier(&self) -> usize {
+        match self {
+            ByteUnit::Byte => 1,
+            ByteUnit::KiB => 1 << 10,
+            ByteUnit::MiB => 1 << 20,
+            ByteUnit::GiB => 1 << 30,
+            ByteUnit::TiB => 1 << 40,
+        }
+    }
+}
```
Comment (Weijun-H) [inline datafusion-cli/src/main.rs:275]: Nice refactor 👍

File: datafusion-cli/src/main.rs [hunk 3]
```diff
@@ -252,26 +253,128 @@ fn is_valid_memory_pool_size(size: &str) -> Result<(), String> {
...
+            return Err(format!(
+                "Negative memory pool size value is not allowed '{}'",
+                size
+            ));
         }
+
+        let num = num_str.parse::<usize>().map_err(|_| {
+            format!("Invalid numeric value in memory pool size '{}'", size)
+        })?;
+
```
Comment (Weijun-H) [inline datafusion-cli/src/main.rs:313]: Can we simplify this? If `num` is of type `usize`, then it is guaranteed to be a positive value.

**Review by Weijun-H — CHANGES_REQUESTED** [review]

File: datafusion-cli/src/main.rs (hunk 3 above)
Comment (yjshen) [reply datafusion-cli/src/main.rs:313]: We need to guard against larger values than usize::MAX?

Comment (Weijun-H) [reply datafusion-cli/src/main.rs:313]:
```suggestion
        let num = num_str.parse::<usize>().map_err(|_| {
            format!("Negative memory pool size value is not allowed '{}'", size)
        })?;
```
Sorry for the confusion, could we test for negative numbers like this?

File: datafusion-cli/src/main.rs (hunk 1 above)
Comment (yjshen) [reply datafusion-cli/src/main.rs:377]: I'm not sure if I understand you correctly. Do you mean the case "-100" shouldn't fail with the Err `"Negative memory pool size value is not allowed '-100'"`?

Comment (Weijun-H) [reply datafusion-cli/src/main.rs:377]: No, we should add a test for cases like `12k12k `.

Comment (yjshen) [reply datafusion-cli/src/main.rs:374]: Nice catch! Fixed in the latest commit.

File: datafusion-cli/src/main.rs (hunk 3 above)
Comment (yjshen) [reply datafusion-cli/src/main.rs]: I have no strong opinion on this.
I changed the code above to parse numeric directly to `usize` for simplicity as suggested, but the caveat is the new error message is less informative, which I suppose should be fine in this case.

**Review by Weijun-H — APPROVED** [review]
LGTM 👍

**Review by alamb — APPROVED** [review]
Looks really nice to me -- thank you @yjshen  and @Weijun-H  ❤️

**Pull request merged**